File: leetcode/p0881/Solution.java
class Solution {
    public int numRescueBoats(int[] people, int limit) {
        Arrays.sort(people);
        
        Queue<Integer> pq = new PriorityQueue<>();
        for (int i = people.length - 1; i >= 0; --i) {
            int p = people[i];
            if (pq.isEmpty()) {
                pq.offer(p);
            } else {
                int min = pq.peek();
                if (p + min <= limit) {
                    pq.poll();
                    pq.offer(limit); // limit 2 people
                } else {
                    pq.offer(p);
                }
            }
        }
        return pq.size();
    }
}

File: leetcode/p0881/Solution.kt
class Solution {
    fun numRescueBoats(people: IntArray, limit: Int): Int {
        people.sort()
        //println(people.toList())
        var count = 0
        var left = 0
        var right = people.size - 1
        while (left <= right) {
            if (people[right] + people[left] > limit) {
                count++
                right--
            } else if (left == right || people[right] + people[left] <= limit) {
                count++
                left++
                right--
            }
        }
        return count
    }
}
